### ISSUE TITLE
[PRE REVIEW]: Peacock.jl: Photonic crystals in Julia

### ISSUE THREAD

**Issue by whedon** [openjournals/joss-reviews#2239]
**Submitting author:** @sp94 (<a href="http://orcid.org/0000-0003-0485-7047">Samuel John Palmer</a>)
**Repository:** <a href="https://github.com/sp94/Peacock.jl" target ="_blank">https://github.com/sp94/Peacock.jl</a>
**Version:** v0.1.0
**Editor:** @dpsanders
**Reviewers:** mlxd, @nmoran
**Managing EiC:** Arfon Smith

**:warning: JOSS reduced service mode :warning:**

Due to the challenges of the COVID-19 pandemic, JOSS is currently operating in a "reduced service mode". You can read more about what that means in [our blog post](https://blog.joss.theoj.org/2020/05/reopening-joss).

**Author instructions**

Thanks for submitting your paper to JOSS @sp94. **Currently, there isn't an JOSS editor assigned** to your paper.

@sp94 if you have any suggestions for potential reviewers then please mention them here in this thread (without tagging them with an @). In addition, [this list of people](https://bit.ly/joss-reviewers) have already agreed to review for JOSS and may be suitable for this submission (please start at the bottom of the list).

**Editor instructions**

The JOSS submission bot @whedon is here to help you find and assign reviewers and start the main review. To find out what @whedon can do for you type:

```
@whedon commands
```

**Comment by whedon** [comment]
Hello human, I'm @whedon, a robot that can help you with some common editorial tasks.

**:warning: JOSS reduced service mode :warning:**

Due to the challenges of the COVID-19 pandemic, JOSS is currently operating in a "reduced service mode". You can read more about what that means in [our blog post](https://blog.joss.theoj.org/2020/05/reopening-joss).

For a list of things I can do to help you, just type:

```
@whedon commands
```

For example, to regenerate the paper pdf after making changes in the paper's md or bib files, type:

```
@whedon generate pdf
```

**Comment by whedon** [comment]
```
Software report (experimental):

github.com/AlDanial/cloc v 1.84  T=0.03 s (468.1 files/s, 51135.1 lines/s)
-------------------------------------------------------------------------------
Language                     files          blank        comment           code
-------------------------------------------------------------------------------
Julia                           11            168             55            923
Markdown                         2             70              0            215
TOML                             2             39              1            147
TeX                              1              5              0            125
-------------------------------------------------------------------------------
SUM:                            16            282             56           1410
-------------------------------------------------------------------------------


Statistical information for the repository '2239' was gathered on 2020/05/23.
No commited files with the specified extensions were found.
```

**Comment by whedon** [comment]
```
Reference check summary:

OK DOIs

- None

MISSING DOIs

- https://doi.org/10.1073/pnas.2133313100 may be missing for title: Coloration strategies in peacock feathers
- https://doi.org/10.1007/978-1-4615-1963-8_41 may be missing for title: Inhibited spontaneous emission in solid-state physics and electronics
- https://doi.org/10.1103/physrevlett.58.2486 may be missing for title: Strong localization of photons in certain disordered dielectric superlattices
- https://doi.org/10.1007/3-540-27701-3_14 may be missing for title: Photonic crystal fibres
- https://doi.org/10.1126/science.1133734 may be missing for title: Quantum spin Hall effect and topological phase transition in HgTe quantum wells
- https://doi.org/10.1103/physrevlett.114.223901 may be missing for title: Scheme for achieving a topological photonic crystal by using dielectric material
- https://doi.org/10.1103/physrevresearch.1.032005 may be missing for title: Engineering fragile topology in photonic crystals: Topological quantum chemistry of light
- https://doi.org/10.1002/qute.201900117 may be missing for title: Tutorial: computing topological invariants in 2D photonic crystals

INVALID DOIs

- None
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.02239/joss.02239/10.21105.joss.02239.pdf)

**Comment by arfon** [comment]
👋 @sp94 - Thanks for your submission to JOSS. As described in [our blog post](https://blog.joss.theoj.org/2020/05/reopening-joss) announcing the reopening of JOSS, we're currently working in a "reduced service mode", limiting the number of papers assigned to any individual editor.

Since reopening JOSS earlier in the week we've had > 50 papers submitted and as such, yours has been put in our backlog that we will be working through over the coming weeks and months.

Thanks in advance for your patience!

**Comment by sp94** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.02239/joss.02239/10.21105.joss.02239.pdf)

**Comment by sp94** [comment]
Thanks for adding me to the waiting list!

From the list of possible reviewers, here are some in Physics/optics/solid state/condensed matter theory, and who also lists Julia in their programming languages:
- crstnbr
- mdavezac
- frssp
- solanpaa

**Comment by kyleniemeyer** [comment]
Hi @jgostick, can you take on this submission as your first editorial assignment?

**Comment by kyleniemeyer** [comment]
@whedon invite @jgostick as editor

**Comment by whedon** [comment]
@jgostick has been invited to edit this submission.

**Comment by jgostick** [comment]
@kyleniemeyer Does it matter that I don't know what a photonic crystal is, have not yet started to use Julia, and can't think of single person who could review this?

**Comment by kyleniemeyer** [comment]
@jgostick I can find one that's a bit closer to your areas of expertise for your first one.

@dpsanders do you have the bandwidth to edit this submission?

**Comment by kyleniemeyer** [comment]
@whedon invite @dpsanders as editor

**Comment by whedon** [comment]
@dpsanders has been invited to edit this submission.

**Comment by dpsanders** [comment]
👋 @kyleniemeyer Yes, but the process will be slow for the next couple of weeks.

**Comment by kyleniemeyer** [comment]
@dpsanders totally understand, I'm in the same situation (hence your invitation coming over the weekend rather than at the beginning of the week...)

**Comment by kyleniemeyer** [comment]
@whedon assign @jgostick as editor

**Comment by whedon** [comment]
OK, the editor is @jgostick

**Comment by kyleniemeyer** [comment]
whoops! sorry @jgostick, that was a mistake.

**Comment by kyleniemeyer** [comment]
@whedon assign @dpsanders as editor

**Comment by whedon** [comment]
OK, the editor is @dpsanders

**Comment by dpsanders** [comment]
👋 Hi @sp94! Thanks for submitting to JOSS and sorry for the delay.

This looks like a very interesting submission. The first thing I notice browsing through is that there are basically no docs. Could you please write some narrative and API documentation? You can use e.g. Documenter.jl for this. 

Also the paper is quite short compared to most JOSS papers. Could you please review the requirements, such as including comparisons to other software and stating the research need?

Thanks!

**Comment by dpsanders** [comment]
@whedon commands

**Comment by whedon** [comment]
Here are some things you can ask me to do:

```
# List all of Whedon's capabilities
@whedon commands

# Assign a GitHub user as the sole reviewer of this submission
@whedon assign @username as reviewer

# Add a GitHub user to the reviewers of this submission
@whedon add @username as reviewer

# Re-invite a reviewer (if they can't update checklists)
@whedon re-invite @username as reviewer

# Remove a GitHub user from the reviewers of this submission
@whedon remove @username as reviewer

# List of editor GitHub usernames
@whedon list editors

# List of reviewers together with programming language preferences and domain expertise
@whedon list reviewers

# Change editorial assignment
@whedon assign @username as editor

# Set the software archive DOI at the top of the issue e.g.
@whedon set 10.0000/zenodo.00000 as archive

# Set the software version at the top of the issue e.g.
@whedon set v1.0.1 as version

# Open the review issue
@whedon start review

EDITORIAL TASKS

# Compile the paper
@whedon generate pdf

# Compile the paper from alternative branch
@whedon generate pdf from branch custom-branch-name

# Remind an author or reviewer to return to a review after a
# certain period of time (supported units days and weeks)
@whedon remind @reviewer in 2 weeks

# Ask Whedon to do a  dry run of accepting the paper and depositing with Crossref
@whedon accept

# Ask Whedon to check the references for missing DOIs
@whedon check references

# Ask Whedon to check repository statistics for the submitted software
@whedon check repository

EiC TASKS

# Invite an editor to edit a submission (sending them an email)
@whedon invite @editor as editor

# Reject a paper
@whedon reject

# Withdraw a paper
@whedon withdraw

# Ask Whedon to actually accept the paper and deposit with Crossref
@whedon accept deposit=true

```

**Comment by dpsanders** [comment]
Hi @sp94! Sorry, I've been out of action for a while but am now picking up again.

Could you please add a couple of examples to the paper too? And briefly comment what a photonic crystal is.

Are photonic crystals in 3D also relevant? How hard would it be to extend the package to that case? Is that in the pipeline? 

For the documentation aspect, [Documenter.jl](https://github.com/JuliaDocs/Documenter.jl) can automatically harvest the docstrings and put them together to make the API docs. I recommend that you expand what you have in the README into the narrative docs. As a useful guide to the different types of documentation I recommend https://documentation.divio.com/.

Thanks!

**Comment by sp94** [comment]
Hi @dpsanders, thanks for taking the time to look at this! I will work through your comments and get back to you this week.

**Comment by sp94** [comment]
Hi @dpsanders,

I just want to update you with where I'm at.

Thanks for recommending Documenter.jl and the divio guide to documentation, they were very helpful. I've ported the old README over to the docs, and split it up into intro/tutorial/how-to's/reference. There are a couple of docstrings still to fix in the reference, and I need to write the "contributing" statement, but I think everything else is ready.

> Also the paper is quite short compared to most JOSS papers. Could you please review the requirements, such as including comparisons to other software and stating the research need?
> Could you please add a couple of examples to the paper too? And briefly comment what a photonic crystal is.
OK, I'll be working on this today/tomorrow.

> Are photonic crystals in 3D also relevant? How hard would it be to extend the package to that case? Is that in the pipeline?
3D is relevant and important, so I have added "2D only" to the README. However, 2D can still be useful for a first approximation of a 3D crystal, and there is still more to learn about 2D crystals with "topology". The method itself can be extended quite easily to 3D, really I think the difficulty is how to keep the interface simple for both 2D and 3D. I will open a GitHub issue with some options for this.

I will let you know in the next few days after I have
(1) completed the few remaining docstrings
(2) reviewed the paper requirements, and added examples to the paper
(3) opened a GitHub issue for extending to 3D

**Comment by dpsanders** [comment]
The docs are looking great!

**Comment by sp94** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.02239/joss.02239/10.21105.joss.02239.pdf)

**Comment by sp94** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.02239/joss.02239/10.21105.joss.02239.pdf)

**Comment by sp94** [comment]
Hi @dpsanders, please let me know if the paper addresses your comments about adding examples, research need, and comparisons to other software

**Comment by dpsanders** [comment]
👋 @sp94: Sorry for the delay here. I think it would be good to include some simple code in the paper showing how to produce one of the figures. You can use one of the examples from the readme.

Please have a look at the [new submission requirements](https://joss.readthedocs.io/en/latest/submitting.html). In particular, there is now a requirement for an explicit "statement of need".

In the paper it says that the package is to "analyse and predict the behaviour". Can you expand there on exactly what the package is able to calculate?

Thanks!

**Comment by dpsanders** [comment]
Also test coverage seems rather low. I suggest you use `@testset`s. You can use an automated service like `coveralls.io` to check test coverage.

**Comment by sp94** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.02239/joss.02239/10.21105.joss.02239.pdf)

**Comment by sp94** [comment]
Hi @dpsanders ,

Sorry for the delay. I've just generated the new PDF which has some example code, the explicit statement of need, and expands on what the package can calculate.

I've linked `coveralls.io` and will get back to you in a couple more days once I have also added the `@testsets`.

Thanks for your comments!

**Comment by sp94** [comment]
Hi @dpsanders,

I've added the test sets and according to coveralls.io I'm at 62% coverage. The remaining percentage is from the plotting functions. Do you have any recommendations in this case?

Thanks for your time

**Comment by sp94** [comment]
Hi @dpsanders, have you had a chance to look at these changes?

**Comment by dpsanders** [comment]
👋  @sp94: Sorry for the delay. Everything is looking good to me, thanks!

I'll start looking for reviewers.

**Comment by dpsanders** [comment]
Sorry for the delay here.

👋 Hi @crstnbr, @mdavezac and @mlxd: Would you be willing to review this submission to JOSS?

**Comment by mlxd** [comment]
Hi @dpsanders I'm happy to review this work.

**Comment by mdavezac** [comment]
@dpsanders, I'm happy to give it go.

**Comment by sp94** [comment]
Thanks for offering your time, everyone

@dpsanders I just noticed that @mdavezac and I are both at Imperial College London, so there could be a perceived conflict of interest. I don't think we've ever met or interacted before so hopefully this can be waived?

**Comment by dpsanders** [comment]
Ah sorry for missing that. Let me consult

**Comment by kyleniemeyer** [comment]
@dpsanders @sp94 Thanks for being upfront about that institutional connection—I think this is fine, though, since you have not actually worked together.

**Comment by danielskatz** [comment]
Let me disagree - this is clearly a conflict, and we can waive it, but if there's another option where we don't have to, it would be better - that should be our first choice.

**Comment by mdavezac** [comment]
As @sp94 said, I don't think we've interacted in any way either. If needed, I am still open to reviewing this paper.

**Comment by dpsanders** [comment]
Thanks @mdavezac and apologies for the situation.

:wave: @datseris and @nmoran: Would you be able to review this submission to JOSS?

**Comment by Datseris** [comment]
Hi @dpsanders , thanks for letting me know. If you had asked one year in the past, I would be happy to review this. But unfortunately at this point in time I neither work in condensed matter, but I am also in a period of preparing several papers in parallel and lack the bandwidth to review this properly :(

**Comment by nmoran** [comment]
Hi @dpsanders, I'm happy to take a look at reviewing this work. It would be my first review for JOSS so wondering if there are some reviewing guidelines or example reviews I could look at?

**Comment by kyleniemeyer** [comment]
@nmoran just jumping in here: we do have explicit [review criteria](https://joss.readthedocs.io/en/latest/review_criteria.html), and in fact when the actual review begins (in a new GitHub issue), you will have a checklist to work through. 

Thanks for your willingness to review this submission!

**Comment by dpsanders** [comment]
Thanks @nmoran! And thanks @kyleniemeyer for the pointer.

If you browse through recent [papers published by JOSS](https://joss.theoj.org), each of them links to the open review, so you can get an idea of how they are carried out.

**Comment by dpsanders** [comment]
@whedon add mlxd as reviewer

**Comment by whedon** [comment]
OK, mlxd is now a reviewer

**Comment by dpsanders** [comment]
@whedon add @nmoran as reviewer

**Comment by dpsanders** [comment]
@whedon start review

**Comment by whedon** [comment]
OK, @nmoran is now a reviewer

**Comment by whedon** [comment]
OK, I've started the review over in https://github.com/openjournals/joss-reviews/issues/2678.

**Comment by nmoran** [comment]
Thanks @dpsanders  and @kyleniemeyer
